users.js

```diff
@@ -0,0 +1,208 @@
+  tableRow.appendChild(deleteTd);
+
+  tableBody.appendChild(tableRow);
+
+  trMap[user.id] = {
+    name: nameTd,
+    email: emailTd,
+    age: ageTd,
+    address: addressTd,
+    row: tableRow,
+  };
+};
+
+const addUserToJSON = (user) => {
+  let database = getDb();
+  let users = database.users;
+  users.push(user);
+  localStorage.setItem("database", JSON.stringify({ ...database, users }));
+};
+
+getUsers().then((usersObj) => {
+  usersObj.forEach(addUserToTable);
+});
+
+let form = document.getElementById("form");
+
+form.addEventListener("submit", (ev) => {
+  ev.preventDefault();
+
+  let id = ev.target.id.value;
+
+  let newId = (Math.random() * 10).toString().replace(".", "").substr(0, 10);
+  let name = ev.target.name.value;
+  let email = ev.target.email.value;
+  let age = ev.target.age.value;
+  let address = ev.target.address.value;
+
+  if (!nameRegex.test(name)) {
+    alert(
+      "Name should be between 1-30 characters, uppercase and lowercase letters and space are allowed."
+    );
+  } else if (!emailRegex.test(email)) {
+    alert("Email should be valid.");
+  } else if (age < 0 || age > 150) {
+    alert("Age should be between 0-150.");
+  } else if (!addressRegex.test(address)) {
+    alert(
+      "Address should be betweeen 1-300 characters and uppercase, lowercase, numbers, space are allowed."
+    );
+  } else {
+    let user = { id: id || newId, name, email, age, address };
+
+    if (id) {
+      editUser(user);
+    } else {
+      addUserToTable(user);
+      addUserToJSON(user);
+    }
+
+    form.reset();
+  }
+});
+
+const loadUserToForm = (user) => {
+  //Loads the data from the local storage, because the form doesn't update until the page is refreshed.
+  let userData = getDb().users.find((u) => u.id === user.id);
+
+  form.id.value = userData.id;
+  form.name.value = userData.name;
+  form.email.value = userData.email;
+  form.age.value = userData.age;
+  form.address.value = userData.address;
+};
+
+const editUser = (user) => {
+  let database = getDb();
+  let users = database.users;
+
+  let userIndex = users.findIndex((u) => {
+    return user.id === u.id;
+  });
+
+  if (userIndex > -1) {
+    users[userIndex] = user;
+    localStorage.setItem("database", JSON.stringify({ ...database, users }));
+    const userObject = trMap[user.id];
+
+    userObject.name.innerHTML = user.name;
+    userObject.email.innerHTML = user.email;
+    userObject.age.innerHTML = user.age;
+    userObject.address.innerHTML = user.address;
+  } else {
+    alert("No user found.");
+  }
+};
+
+const getDb = () => {
+  return JSON.parse(localStorage.getItem("database"));
+};
+
+const deleteUser = (user) => {
+  let deleteModal = confirm(`Do you want to delete user ${user.name}?`);
+
+  if (deleteModal) {
+    tableBody.removeChild(trMap[user.id].row);
+
+    let database = getDb();
+    let users = database.users.filter((u) => u.id !== user.id);
+    localStorage.setItem("database", JSON.stringify({ ...database, users }));
```

Comment:
Do you find any repeating pattern in these calls? Try to avoid duplications. Think for abstractions which will allow you to avoid them.